src/commands/project/append-variable-group.ts

```diff
@@ -0,0 +1,175 @@
+import { SERVICE_PIPELINE_FILENAME } from "../../lib/constants";
+
+// Values that need to be pulled out from the command operator
+export interface CommandOptions {
+  orgName: string | undefined;
+  personalAccessToken: string | undefined;
+  devopsProject: string | undefined;
+}
+
+// Configuration values
+interface ConfigValues {
+  orgName: string;
+  personalAccessToken: string;
+  devopsProject: string;
+}
+
+/**
+ * Validates values passed as options
+ * @param opts The initialized options
+ */
+export const validateValues = (opts: CommandOptions): ConfigValues => {
+  populateInheritValueFromConfig(decorator, Config(), opts);
+  validateForRequiredValues(decorator, opts, true);
+
+  // validateForRequiredValues already check required values
+  // || "" is just to satisfy eslint rule.
+  validateProjectNameThrowable(opts.devopsProject || "");
+  validateOrgNameThrowable(opts.orgName || "");
+
+  return {
+    orgName: opts.orgName || "",
+    personalAccessToken: opts.personalAccessToken || "",
+    devopsProject: opts.devopsProject || "",
+  };
+};
+
+/**
+ * Check project dependencies
+ * @param projectPath Path to the project directory
+ */
+export const checkDependencies = (projectPath: string): void => {
+  const fileInfo: BedrockFileInfo = bedrockYaml.fileInfo(projectPath);
+  if (fileInfo.exist === false) {
+    throw buildError(
+      errorStatusCode.VALIDATION_ERR,
+      "project-append-variable-group-cmd-err-dependency"
+    );
+  }
+};
+
+/**
+ * Checks if a variable group exists in an Azure DevOps project
+ * @param variableGroupName Variable Group Name
+ * @param values The config values
+ */
+export const variableGroupExists = async (
+  variableGroupName: string,
+  values: ConfigValues
+): Promise<boolean> => {
+  const accessOpts: AzureDevOpsOpts = {
+    orgName: values.orgName,
+    personalAccessToken: values.personalAccessToken,
+    project: values.devopsProject,
+  };
+
+  return await hasVariableGroup(accessOpts, variableGroupName);
+};
+
+/**
+ * Update the variable groups for the project services
+ * @param bedrockFile The bedrock.yaml file
+ * @param variableGroupName Variable Group Name
+ */
+export const updateServicesVariableGroups = (
+  bedrockFile: BedrockFile,
+  variableGroupName: string
+): void => {
+  bedrockFile.services.forEach((service) => {
+    const path = service.path;
+    appendVariableGroupToPipelineYaml(
+      path,
+      SERVICE_PIPELINE_FILENAME,
+      variableGroupName
+    );
+  });
+};
+
+/**
+ * Executes the command.
+ *
+ * @param projectPath The path to the spk project
+ * @param variableGroupName Variable Group Name
+ * @param opts Option object from command
+ * @param exitFn The exit function
+ */
+export const execute = async (
+  projectPath: string,
+  variableGroupName: string,
+  opts: CommandOptions,
+  exitFn: (status: number) => Promise<void>
+): Promise<void> => {
+  if (!hasValue(variableGroupName)) {
+    await exitFn(1);
+    return;
+  }
+
+  try {
```

Comment:
please have the hasValue check in try block, so that we have the error chaining.
```
try {
  if (!hasValue(variableGroupName)) {
    throw buildError(...
  }

```